### PR TITLE
[Intents] Add bindings for Xcode 9 Beta 1, 2, 3, 4 & 5

### PR DESCRIPTION
**Pull request by dalexsoto**

### CODE REVIEW

File: src/intents.cs [hunk 1]
```diff
@@ -1579,14 +2038,31 @@ interface INCurrencyAmountResolutionResult {
...
 		NSDateComponents EndDateComponents { get; }
+
+		[Watch (4,0), Mac (10,13, onlyOn64:true), iOS (11,0)]
+		[NullAllowed, Export ("recurrenceRule", ArgumentSemantic.Copy)]
+		INRecurrenceRule RecurrenceRule { get; }
+
+		[Watch (4,0), Mac (10,13, onlyOn64:true), iOS (11,0)]
+		[Export ("EKRecurrenceRule")]
+		[return: NullAllowed]
+		EKRecurrenceRule GetEKRecurrenceRule ();
```
Comment (rolfbjarne) [inline src/intents.cs]: Can this be a readonly property instead?

File: src/intents.cs [hunk 2]
```diff
@@ -2364,13 +2883,25 @@ interface INListRideOptionsIntentResponse {
 	[DisableDefaultCtor]
 	interface INMessage : NSCopying, NSSecureCoding {
 
-		[Export ("initWithIdentifier:content:dateSent:sender:recipients:")]
+		[Watch (4,0), Mac (10,13, onlyOn64:true), iOS (11,0)]
+		[Export ("initWithIdentifier:conversationIdentifier:content:dateSent:sender:recipients:groupName:messageType:")]
```
Comment (rolfbjarne) [inline src/intents.cs]: The designated initializer changed?

File: src/intents.cs [hunk 3]
```diff
@@ -2294,6 +2806,13 @@ interface INInteraction : NSSecureCoding, NSCopying {
 
 		[NullAllowed, Export ("groupIdentifier")]
 		string GroupIdentifier { get; set; }
+
+		// From INParameter.h INInteraction ()
+
+		[iOS (11,0), Watch (4,0), NoMac]
+		[Export ("parameterValueForParameter:")]
+		[return: NullAllowed]
+		NSObject GetParameterValue (INParameter parameter);
```
Comment (spouliot) [inline src/intents.cs]: I wonder if this should be made `[Internal]` and expose an API with a generic type ?
This looks nicer than a cast but there's also the case where `(id)` can sometime be an `INativeObject` and not an `NSObject` subclass

File: src/intents.cs (hunk 2 above)
Comment (spouliot) [inline src/intents.cs]: that make sense, the new API has an extra argument so it likely calls the _old designated_ init, which makes it non-designated for subclassing.

Comment (dalexsoto) [inline src/intents.cs]: Yup it did

```objc
- (instancetype)initWithIdentifier:(NSString *)identifier
            conversationIdentifier:(nullable NSString *)conversationIdentifier
                           content:(nullable NSString *)content
                          dateSent:(nullable NSDate *)dateSent
                            sender:(nullable INPerson *)sender
                        recipients:(nullable NSArray<INPerson *> *)recipients
                         groupName:(nullable INSpeakableString *)groupName
                       messageType:(INMessageType)messageType NS_DESIGNATED_INITIALIZER API_AVAILABLE(macosx(10.13), ios(11.0), watchos(4.0));
```

File: src/intents.cs (hunk 1 above)
Comment (dalexsoto) [inline src/intents.cs]: Yeah just went to docs again and I do not see any reason not to do so.

File: src/intents.cs (hunk 3 above)
Comment (dalexsoto) [inline src/intents.cs]: We can constrain it to `NSObject` for now and if needed expand it to `INativeObject` in the future

Comment (spouliot) [inline src/intents.cs]: That's _generally_ a breaking change and not a good candidate for overload (in C#) since only the return type is different

Comment (dalexsoto) [inline src/intents.cs]: Not really because you will need to specify the `bool owns` parameter for `NativeObject`

Comment (dalexsoto) [inline src/intents.cs]: Unless we want to hide the ownership here for `INativeObject`

**Comment by dalexsoto** [comment]
@rolfbjarne feedback implemented 👍

**Pull request merged**